### ISSUE TITLE
Default rbd pool

### ISSUE THREAD

**Issue by sajal** [rook/rook#554]
I am playing with kubernetes (and rook) for potential production deployment in the future. I am toying with ~5 node cluster.

I see rook(or ceph) creates a `rbd` pool. My understanding is that this pool is not used for anything.

I always see warning like `too many PGs per OSD (790 > max 300)` . 

Recently I did an unclean removal of a node (and thus killing the OSD) . And now have a lot of warnings and errors.

```
# ceph -s
    cluster 328151bf-770b-4d58-80c6-99fd003cbb32
     health HEALTH_ERR
            346 pgs are stuck inactive for more than 300 seconds
            345 pgs stale
            1 pgs stuck inactive
            345 pgs stuck stale
            1 pgs stuck unclean
            too many PGs per OSD (790 > max 300)
     monmap e3: 2 mons at {mon0=10.2.99.5:6790/0,mon1=10.2.41.4:6790/0}
            election epoch 6, quorum 0,1 mon1,mon0
        mgr no daemons active 
     osdmap e64: 5 osds: 5 up, 5 in; 295 remapped pgs
            flags sortbitwise,require_jewel_osds,require_kraken_osds
      pgmap v107096: 2148 pgs, 2 pools, 3642 MB data, 1044 objects
            75479 MB used, 302 GB / 376 GB avail
                1802 active+clean
                 345 stale+active+clean
                   1 creating
  client io 299 kB/s wr, 0 op/s rd, 135 op/s wr
```

My thinking is that the errors relate to the unused `rbd` pool, since the pool I created using TPR has replication.count of 2 and everything works fine.

Can we :-

1. Suppress warnings/errors of pool we don't use
2. Not create a pool that nobody will use
3. Tune properties of default pool.

FYI I'm using `v0.3.1`

**Comment by bzub** [comment]
You can tune the lower level settings like default pool size, PGs size, etc using `rookd`'s `--ceph-config-override` argument.  Settings in that file will be merged into the rook config that daemons use.  I use a configmap with my ceph.conf settings inside, and modified the rook OSD DaemonSet to use those settings.  Also, recently Rook gained the ability to specify some of those pool settings from the start in a [Storage Pool TPR](https://github.com/rook/rook/blob/master/Documentation/pool-tpr.md)

You can also change pool/OSD settings on a running cluster using the rook-tools pod and using the ceph/rbd commands directly.

The RBD pool is pretty standard, useful for getting started. I use it for my "slow" disks and created another pool for SSD drives.  That's all completely up to the user.  For now the number of PGs is not dynamically chosen so with 5 OSDs you can recreate/create a pool with fewer PGs.  You can delete the pool if you like if you don't use it.  The only trick is enabling the deletion of pools.  Here's how to do it on the rook-tools pod:

##  Create rook-tools pod
```bash
kubectl -n rook create -f https://github.com/rook/rook/blob/master/toolbox/rook-tools.yml
```
 
## Enter a shell inside rook-tools, enable pool deletion, delete rbd pool
kubectl -n rook exec -it rook-tools -- bash
ceph tell mon.* injectargs -- --mon-allow-pool-delete
ceph osd pool rm rbd rbd --yes-i-really-really-mean-it  
```

Hope this helps.  You're right that there is more room for improvement on default pool settings, however with the ceph override option it's at least possible to set those defaults on a cluster manually, for now.

**Comment by sajal** [comment]
@bzub Thanks for the info, I will play with tuning things later, once I get some better grasp of how ceph works.

The instruction about deleting the pool was very helpful. I did not know one could simply `tell` the mon to change the configuration. I now have a shiny storage cluster without any error or warnings.

**Comment by kokhang** [comment]
@sajal I think the fix is to not create that unused pool. I think Ceph is creating it by default. Nice find. Thanks @bzub for the clear steps to workaround this issue.

**Comment by jbw976** [comment]
I think I understand this issue better now.  First, the "unclean removal of a node (and thus killing the OSD)" had more of an effect on the health of PGs than potentially was alluded to earlier in this thread.  That being said, the default rbd pool that Ceph (not Rook) creates does have more of an impact that I previously expected.

This default rbd pool gets created with a whopping 2048 placement groups, which depending on your cluster size (number of OSDs) is likely too big.  In our simple development clusters, this is always too big and we always see the `too many PGs per OSD` warning.  What's weird is that Rook by default sets the following config so that by default storage pools will get 100 PGs, but that doesn't seem to be honored by Ceph when it creates the default rbd pool:
```
osd pool default pg num  = 100
osd pool default pgp num = 100
```

I can't find anyway to prevent Ceph from automatically creating this pool, so it would seem that we'd need to delete it in the rook-operator, perhaps after mons have formed quorum but before we have deployed any OSDs.  This deletion would need to be intelligent (idempotent) though in the case that the cluster has already been set up, we would NOT want to delete the rbd pool if there is any data in it!

The advantage of the pool though is that all the tools use it as a default: `ceph`, `rbd`, and `rook`.  It is nice having a default pool setup automatically so the user doesn't ever have to create a pool or specify a pool in any subsequent commands such as creating a block image.

So perhaps the rbd default pool is nice to have, but I'd like to better understand why it uses 2048 PGs and not the default of 100 that is specified in the ceph config that rook generates. Anyone know why that is?

**Comment by bassam** [comment]
It's because of https://github.com/rook/rook/blob/master/pkg/cephmgr/mon/config.go#L299

**Comment by jbw976** [comment]
Ah, I see where in the code the `osd pg bits` config value would affect the number of placement groups for the default RBD pool (1 << 11 == 2048):
https://github.com/rook/ceph/blob/rook-luminous/src/osd/OSDMap.cc#L2702
https://github.com/rook/ceph/blob/rook-luminous/src/osd/OSDMap.cc#L2734

My understanding is that this config value represents the number of bits to use to store the maximum number of PGs per OSD and since we specify 11 then that max is 2048 for us.  But I don't understand why that max gets used in the code above for the number of PGs in the default RBD pool instead of the `osd pool default pg num` config value.  That's odd to me.

**Comment by xeor** [comment]
Will there be a negative effect if you get that warning? I'm getting the same thing. Guess everyone with less than 8 or so nodes will have​that warning?

**Comment by jbw976** [comment]
It is suboptimal to have too many placement groups assigned to each OSD, but full functionality will still be present.  According to the Ceph docs, the OSDs will require more resources and it will take them more time to peer, but they will still be 100% functional:
http://docs.ceph.com/docs/master/rados/operations/placement-groups/#choosing-the-number-of-placement-groups

**Comment by jbw976** [comment]
This is actually fixed with Rook v0.5.0 that is using Ceph Luminous by default.  The `rbd` pool is no longer created by default.
